Adds guidelines and examples for code contribution (#1876)
isaac-sim/IsaacLab: docs/source/refs/contributing.rst

  * ``Fixed``: For any bug fixes.

* Each change is described in its corresponding sub-section with a bullet point.

> * The bullet points are written in the **past tense and in imperative mode**.

Reviewer: This is a bit confusing - past tense would be what you have below "Added ____", "Fixed _____". What does the imperative mode add to this? I thought imperative mood (NIT: it's mood, not mode) would be "Add ___`, "Fix ___"

If we want the first one, I think past tense is more clear
```suggestion
* The bullet points are written in the **past tense**.
```

I'm also confused as I thought previously you wanted commits and PR titles to be of the form "Adds ____" rather than "Added ___". Can we stick to the same format for commits, PRs and the CHANGELOG to make things simpler (and therefore easier to follow)? I think for people coming from other projects "Added ___" is more common, and hence it can be hard for people to adapt to this new formatting you're requesting

Reviewer: Thanks for the link - I don't see anything about past imperative mood there though. From some googling it doesn't seem to exist in English (and if it does, it's something like "You should have done  X", where X is the action being directed)

This is getting pedantic - only wanted to say that what you are describing likely might be more confusing so I suggest removing the imperative bit

I meant for the final squashed commit that shows up in the commit history, but to me, PR Title, Changelog and commit verb tenses are all meh, but if you want to be strict on them, we should be clear at least! Happy to discuss more offline if needed

it can result in circular imports because, in many configuration objects, we specify classes or functions
as default values using the attributes ``class_type`` and ``func`` respectively.

To address circular imports, we leverage the `typing.TYPE_CHECKING
<https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING>`_ variable. This special variable is
evaluated only during type-checking, allowing us to import classes or functions in the configuration objects
without triggering circular imports.

It is important to note that this is the sole instance within our codebase where circular imports are used
and are acceptable. In all other scenarios, we adhere to best practices and recommend that you do the same.

Type-hinting
^^^^^^^^^^^^

To make the code more readable, we use `type hints <https://docs.python.org/3/library/typing.html>`__ for
all the functions and classes. This helps in understanding the code and makes it easier to maintain. Following
this practice also helps in catching bugs early with static type checkers like `mypy <https://mypy.readthedocs.io/en/stable/>`__.

**Type-hinting only in the function signature**

> To avoid duplication of efforts, we do not specify type hints for the arguments and return values in the docstrings.

Reviewer: ```suggestion
To avoid duplications, which can be hard to synchronize when making changes, we do not specify type hints for the arguments and return values in the docstrings.
```

For instance, the following are bad examples for various reasons:

.. code:: python

   def my_function(a, b):

>       """
>       This is a bad example.

Reviewer: ```suggestion
      """This is a bad example.
```

>    def my_function(a, b):
>       """
>       This is a bad example.
> 
>       Reason: No type hints anywhere.

Author: ```suggestion
   def my_function(a, b):
      """Adds two numbers.

      This function is a bad example. Reason: No type hints anywhere.
```

      Args:
         a: The first argument.
         b: The second argument.

      Returns:
         The sum of the two arguments.
      """
      return a + b

.. code:: python

   def my_function(a, b):

>       """
>       This is a bad example.

Reviewer: ```suggestion
      """This is a bad example.
```

>    def my_function(a, b):
>       """
>       This is a bad example.
> 
>       Reason: Type hints in the docstring and not in the function signature.

Author: ```suggestion
   def my_function(a, b):
      """Adds two numbers.

      This function is a bad example. Reason: Type hints in the docstring and not in the
      function signature.
```

      Args:
         a (int): The first argument.
         b (int): The second argument.

      Returns:
         int: The sum of the two arguments.
      """
      return a + b

.. code:: python

   def my_function(a: int, b: int) -> int:

>       """
>       This is a bad example.

Reviewer: ```suggestion
      """This is a bad example.
```

>    def my_function(a: int, b: int) -> int:
>       """
>       This is a bad example.
> 
>       Reason: Type hints in the docstring and in the function signature. Redundancy.
> 

Author: ```suggestion
   def my_function(a: int, b: int) -> int:
      """Adds two numbers.

      This function is a bad example. Reason: Type hints in the docstring and in the function 
      signature. Redundancy.

```

      Args:
         a (int): The first argument.
         b (int): The second argument.

      Returns:
         int: The sum of the two arguments.
      """
      return a + b

The following is how we expect you to write the docstrings and type hints:

.. code:: python

   def my_function(a: int, b: int) -> int:

>       """
>       This is a good example.

Reviewer: The docstrings should start on the first line, this might be confusing to people
```suggestion
      """This is a good example.
```

>       """
>       This is a good example.
> 
>       Reason: Type hints in the function signature and not in the docstring.

Author: ```suggestion
      """Adds two numbers.

      This function is a good example. Reason: Type hints in the function signature and not in the 
      docstring.
```

to keep up with the development pace.

**What is considered as a bad documentation?**

* If someone else wants to use the code, they cannot understand the code just by reading the documentation.

  What this means is that the documentation is not complete or is not written in a way that is easy to understand.
  The next time someone wants to use the code, they will have to spend time understanding the code (in the best
  case scenario), or scrap the code and start from scratch (in the worst case scenario).

* Certain design subtleties are not documented and are only apparent from the code.

  Often certain design decisions are made to address specific use cases. These use cases are not
  obvious to someone who wants to use the code. They may change the code in a way that is not intuitive
  and unintentionally break the code.

* The documentation is not updated when the code is updated.

  This means that the documentation is not kept up to date with the code. It is important to update the
  documentation when the code is updated. This helps in keeping the documentation up to date and in sync
  with the code.

> **What is considered as a good documentation?**

Reviewer: ```suggestion
**What is considered good documentation?**
```
